File: src/theShortestRoute/TheShortestRoute.java
package theShortestRoute;

import java.io.*;
import java.util.*;

public class TheShortestRoute {
    static class Node implements Comparable<Node>{
        int index;
        int value;

        public Node(int index, int value){
            this.index = index;
            this.value = value;
        }

        @Override
        public int compareTo(Node o) {
            if(value > o.value){
                return 1;
            }else if(value < o.value){
                return -1;
            }else{
                return 0;
            }
        }
    }
    public static void main(String[] args) throws IOException {
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String[] ve = br.readLine().split(" ");
        int v = Integer.parseInt(ve[0]), e = Integer.parseInt(ve[1]);
        int start = Integer.parseInt(br.readLine());

        List<List<Node>> graph = new ArrayList<>(v + 1);
        int[] arr = new int[v + 1];

        for (int i = 0; i < v + 1; i++) {
            graph.add(new ArrayList<>());
            arr[i] = Integer.MAX_VALUE;
        }

        arr[start] = 0;

        for (int i = 0; i < e; i++) {
            String[] uvw = br.readLine().split(" ");
            int u = Integer.parseInt(uvw[0]), V = Integer.parseInt(uvw[1]), w = Integer.parseInt(uvw[2]);

            graph.get(u).add(new Node(V, w));
        }

        Queue<Node> queue = new PriorityQueue<>();
        queue.add(new Node(start, 0));

        while(!queue.isEmpty()){
            Node node = queue.remove();
            int index = node.index, value = node.value;

            List<Node> list = graph.get(index);

            if(arr[index] < value) continue;

            for (int i = 0; i < list.size(); i++) {
                Node vw = list.get(i);

                if(arr[vw.index] > arr[index] + vw.value){
                    arr[vw.index] = arr[index] + vw.value;
                    queue.add(new Node(vw.index, arr[vw.index]));
                }
            }
        }

        StringBuilder sb = new StringBuilder();

        for (int i = 1; i < arr.length; i++) {
            if(arr[i] == Integer.MAX_VALUE) sb.append("INF\n");
            else sb.append(String.valueOf(arr[i]) + "\n");
        }

        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }
}

File: JS/theShortestRoute/theShortestRoute.js
"use strict";

const fs = require("fs");

const input =
  process.platform === "linux"
    ? fs.readFileSync("dev/stdin").toString().split("\n")
    : `5 6
1
5 1 1
1 2 2
1 3 3
2 3 4
2 4 5
3 4 6`.split("\n");

class Heap {
  constructor() {
    this.heap = [];
  }

  getLeftChildIndex = (parentIndex) => parentIndex * 2 + 1;
  getRightChildIndex = (parentIndex) => parentIndex * 2 + 2;
  getParentIndex = (childIndex) => Math.floor((childIndex - 1) / 2);

  peek = () => this.heap[0];

  insert = (key, value) => {
    const node = { key, value };
    this.heap.push(node);
    this.heapifyUp();
  };

  heapifyUp = () => {
    let index = this.heap.length - 1;
    const lastInsertedNode = this.heap[index];

    while (index > 0) {
      const parentIndex = this.getParentIndex(index);

      if (this.heap[parentIndex].value > lastInsertedNode.value) {
        this.heap[index] = this.heap[parentIndex];
        index = parentIndex;
      } else break;
    }

    this.heap[index] = lastInsertedNode;
  };

  remove = () => {
    const count = this.heap.length;
    const rootNode = this.heap[0];

    if (count <= 0) return undefined;
    if (count === 1) this.heap = [];
    else {
      this.heap[0] = this.heap.pop();
      this.heapifyDown();
    }

    return rootNode;
  };

  heapifyDown = () => {
    let index = 0;
    const count = this.heap.length;
    const rootNode = this.heap[index];

    while (this.getLeftChildIndex(index) < count) {
      const leftChildIndex = this.getLeftChildIndex(index);
      const rightChildIndex = this.getRightChildIndex(index);

      const smallerChildIndex =
        rightChildIndex < count &&
        this.heap[rightChildIndex].value < this.heap[leftChildIndex].value
          ? rightChildIndex
          : leftChildIndex;

      if (this.heap[smallerChildIndex].value <= rootNode.value) {
        this.heap[index] = this.heap[smallerChildIndex];
        index = smallerChildIndex;
      } else break;

      this.heap[index] = rootNode;
    }
  };
}

class PriorityQueue extends Heap {
  constructor() {
    super();
  }

  enqueue = (priority, value) => this.insert(priority, value);
  dequeue = () => this.remove();
  isEmpty = () => this.heap.length <= 0;
}

const [V, E] = input[0].split(" ").map((v) => parseInt(v));
const start = parseInt(input[1]);

const graph = Array.from(Array(V + 1), () => []);
const arr = Array(V + 1).fill(Infinity);

arr[start] = 0;

for (let i = 2; i < E + 2; i++) {
  const [u, v, w] = input[i].split(" ").map((v) => parseInt(v));

  graph[u].push([v, w]);
}

const pq = new PriorityQueue();
pq.enqueue(start, 0);

while (!pq.isEmpty()) {
  const { key, value } = pq.dequeue();

  if (arr[key] < value) continue;

  for (let i = 0; i < graph[key].length; i++) {
    const [v, w] = graph[key][i];
    if (arr[v] > arr[key] + w) {
      arr[v] = arr[key] + w;
      pq.enqueue(v, arr[v]);
    }
  }
}

const result = [];

for (let i = 1; i < arr.length; i++) {
  if (arr[i] === Infinity) result.push("INF");
  else result.push(arr[i]);
}

console.log(result.join("\n"));
